lib/utility.js

```diff
@@ -1,10 +1,171 @@
+      success: function(res){
+        if (!res.code) reject({"ErrorMessage": "Fail to Generate Login Code"});
+        requestPromise(loginUrl, {"LoginCode": res.code})
+        .then(function(data){
+          app = getApp();
+          if (globalData) globalData.userInfo = data.UserId;
+          if (app) app.globalData.userInfo = data.UserId;
+          return resolve(data.userId);
+         })
+      },
+      fail: function(res){
+        reject({'message': "Fail to Login"});
+      }
+    });
+  })
+}
+
+
+/** @review
+ * Query Server for my product list 
+ * @param offset: request offset
+ * @param offset: request pageSize
+ */
+function getMyProduct(offset, pageSize){
+  return getUserId(getApp().globalData) // Make sure UserId is retrieved
+          .then((id)=>{
+            let data = {
+              "Action": "GetProductsByUserId",
+              "UserId": id
+            };
+            if (offset) data["StartId"] = offset;
+            if (pageSize) data["pageSize"] = pageSize;
+            return requestPromise(marketUrl, data);
+          })
+          .then((res)=>{ return (res.Products.map((x) => new Product(x))); });
+}
+
+/** @review
+ * post 1 product and its images (auxiliary)
+ * @param data.ProductName: String 
+ * @param data.ProductType: Number
+ * @param data.ProductImages: Array[String] (OPTIONAL)
+ * @promise resolve: Server response
+ * @promise reject: Lack of Info
+ */
+function postProduct(data) {
+  // Check validity
+  if (!data.ProductName)
+      throw { 'message': "Lack product name" };
+  if (!data.ProductType)
+      throw { 'message': "Lack product type" };
+  data.ProductOwner = getApp().globalData.userInfo;
+  // Post Product
+  data["Action"] = "PostProduct";
+  let postChain =  requestPromise(marketUrl, data);
+
+  // Post Product Images
+  if (!data.ProductImages) return postChain;
+  // Extract Product Id
+  postChain = postChain.then(res => {data.ProductId = res.ProductId});
+  let promisePointer = 0;
+  let imgConfig = { "Action": "PostProductImages"};
+  for (var i=0;i<data.ProductImages.length;i++){ // post each image
+      postChain = postChain.then(function(res){
+          // get product id asynchronously
+          imgConfig["ProductId"] = data.ProductId;  
+          return requestPromise(marketUrl, imgConfig, data.ProductImages[promisePointer++]);
+      });
+  }
+  return postChain;
+}
+
+/** @review
+ * post multiple products and its images
+ * @param data: Array of Product
+ *        Product.ProductName: String 
+ *        Product.ProductType: Number
+ *        Product.ProductOwner: Number (app.globalData.userInfo)
+ *        Product.ProductImages: Array[String] (OPTIONAL)
+ * @promise resolve: Server response
+ * @promise reject: Lack of Info
+ */
+function postProducts(data) {
+  let productLine = getUserId();
+  let postPointer = 0;
+  for (var i=0;i<data.length;i++){
+      productLine = productLine.then(function(res){
+        return postProduct(data[postPointer++]);

```

Comment:
postPointer is a global variable used in an asynchronous callback. In this scenario, it would always be the data.length.

Basically, this is what's gonna happen.
The for loop runs for data.length-1 iteration, each iteration it increments postPointer and registers a callback, which, when gets run, references postPointer.

When the callbacks are actually called, the for loop has already finished (because it's synchronous) and the postPointer has already been incremented. Then when the callback goes to fetch the value stored in the reference stored in postPointer, it is data.length.